sql/sql_regex.h

```diff
@@ -0,0 +1,132 @@
+/* Copyright (c) 2016, Percona and/or its affiliates. All rights reserved.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA */
+
+#ifndef _regex_list_handler_h_
+#define _regex_list_handler_h_
+
+#include <string>
+#include <algorithm>
+
+#include "my_global.h"
+#include "my_pthread.h"
+#include "my_regex.h"
+
+/*
+  Wrapper class aroung my_regex that manages concurrent acess and the lifetime
+  of the compiled regex.
+*/
+class Regex
+{
+private:
+  std::string m_pattern;
+  my_regex_t m_expr;
+  bool m_compiled;
+
+  mutable mysql_rwlock_t m_rwlock;
+
+  Regex(const Regex& other) {}
+  Regex& operator=(const Regex& other)
+  {
+    return *this;
+  }
+
+public:
+  Regex() :
+    m_pattern(""),
+    m_compiled(false)
+  {
+    mysql_rwlock_init(0, &m_rwlock);
+  }
+
+  ~Regex()
+  {
+    mysql_rwlock_wrlock(&m_rwlock);
+    if (m_compiled)
+    {
+      my_regfree(&m_expr);
+      m_compiled = false;
+    }
+    mysql_rwlock_unlock(&m_rwlock);
+    mysql_rwlock_destroy(&m_rwlock);
+  }
+
+  bool compiled() const
+  {
+    return m_compiled;
+  }
+
+  const std::string& pattern() const
+  {
+    return m_pattern;
+  }
+
+  // Compile the pattern into an expression
+  bool compile(const char *pattern)
+  {
+    int error = 0;
+
+    mysql_rwlock_wrlock(&m_rwlock);
+
+    m_pattern.clear();
+    if (m_compiled)
+    {
+      my_regfree(&m_expr);
+      m_compiled = false;
+    }
+
+    if (pattern)
+    {
+      m_pattern.assign(pattern);
+    }
+
+    if (!m_pattern.empty())
+    {
+      error = my_regcomp(&m_expr,
+                         m_pattern.c_str(),
+                         MY_REG_EXTENDED | MY_REG_NOSUB | MY_REG_ICASE,
```

Comment:
MY_REG_ICASE suggests a subtle bug here: MySQL table name case sensitivity depends on the underlying filesystem and lower_case_table_names value, see https://dev.mysql.com/doc/refman/5.7/en/identifier-case-sensitivity.html.

To resolve, I'd add bool case_sensitive arg to constructor or here (whichever makes more sense), and work out the correct value at the caller. I'd add some MTR too (extend the existing one with lower_case_table_names = 1, and since this depends on FS, add a new testcase with include/have_case_sensitive_file_system.inc for l_c_t_n = 0. For absolute completeness yet another new testcase with include/have_case_insensitive_file_system and l_c_t_n = 2 would also be required, but we don't have a supported platform aka MacOS for that).

Since std::regex is case-sensitive by default, this also suggests that the upstream has the opposite bug here which should be reported to them (option broken if table names case-insensitive)